### PR TITLE
Compatibility with PHP 7 and 8

### PR DESCRIPTION
**Pull request by remicollet**

### CODE REVIEW

File: facedetect.cc [hunk 1]
```diff
@@ -31,12 +31,13 @@ extern "C" {
...
+/* for PHP 8 */
+#ifndef TSRMLS_CC
+#define TSRMLS_CC
+#endif
 
 using namespace cv;
 
-/* True global resources - no need for thread safety here */
-static int le_facedetect;
-
```
Comment (remicollet) [inline facedetect.cc:39]: Unused

File: facedetect.cc [hunk 2]
```diff
@@ -49,13 +50,18 @@ static void php_facedetect(INTERNAL_FUNCTION_PARAMETERS, int return_type) {
     zval *pArray;
 
     char *file = NULL, *casc = NULL;
+#if PHP_VERSION_ID < 70000
     long flen, clen;
+#else
+    size_t flen, clen;
```
Comment (remicollet) [inline facedetect.cc:56]: Mandatory with PHP 7+ to avoid memory corruption

File: facedetect.cc [hunk 3]
```diff
@@ -132,6 +138,27 @@ PHP_INI_ENTRY("facedetect.cascade", "", PHP_INI_ALL, on_cascade_change)
 PHP_INI_END()
 
 
+#if PHP_VERSION_ID < 80000
+ZEND_BEGIN_ARG_INFO_EX(arginfo_face_detect, 0, 0, 1)
+	ZEND_ARG_INFO(0, image_path)
+	ZEND_ARG_INFO(0, cascade_path)
+ZEND_END_ARG_INFO()
+
+#define arginfo_face_count arginfo_face_detect
```
Comment (remicollet) [inline facedetect.cc:147]: Simple arginfo for PPHP 5 and 7

File: facedetect.cc [hunk 4]
```diff
@@ -132,6 +138,27 @@ PHP_INI_ENTRY("facedetect.cascade", "", PHP_INI_ALL, on_cascade_change)
...
+#else
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_MASK_EX(arginfo_face_detect, 0, 1, MAY_BE_FALSE|MAY_BE_ARRAY)
+	ZEND_ARG_TYPE_INFO(0, image_path, IS_STRING, 0)
+	ZEND_ARG_TYPE_INFO_WITH_DEFAULT_VALUE(0, cascade_path, IS_STRING, 1, "null")
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_MASK_EX(arginfo_face_count, 0, 1, MAY_BE_FALSE|MAY_BE_LONG)
+	ZEND_ARG_TYPE_INFO(0, image_path, IS_STRING, 0)
+	ZEND_ARG_TYPE_INFO_WITH_DEFAULT_VALUE(0, cascade_path, IS_STRING, 1, "null")
+ZEND_END_ARG_INFO()
```
Comment (remicollet) [inline facedetect.cc:158]: arginfo are mandatory with PHP 8
Also add type hinting.

File: facedetect.cc [hunk 5]
```diff
@@ -49,13 +50,18 @@ static void php_facedetect(INTERNAL_FUNCTION_PARAMETERS, int return_type) {
...
+    size_t flen, clen;
+#endif
+
 
     Mat img;
     Mat gray;
     std::vector<Rect> faces;
 
-    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|s", &file, &flen, &casc, &clen) == FAILURE) {
+    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "p|p", &file, &flen, &casc, &clen) == FAILURE) {
```
Comment (remicollet) [inline facedetect.cc:64]: "p" is for path, to check for nul byte

**Comment by remicollet** [comment]
```
$ php -n -d extension=modules/facedetect.so --re facedetect
Extension [ <persistent> extension #16 facedetect version 0.1.0 ] {

  - INI {
    Entry [ facedetect.cascade <ALL> ]
      Current = ''
    }
  }

  - Functions {
    Function [ <internal:facedetect> function face_detect ] {

      - Parameters [2] {
        Parameter #0 [ <required> string $image_path ]
        Parameter #1 [ <optional> ?string $cascade_path = null ]
      }
      - Return [ array|false ]
    }
    Function [ <internal:facedetect> function face_count ] {

      - Parameters [2] {
        Parameter #0 [ <required> string $image_path ]
        Parameter #1 [ <optional> ?string $cascade_path = null ]
      }
      - Return [ int|false ]
    }
  }
}

```

**Comment by infusion** [comment]
Thanks for the compat changes!!

**Pull request merged**